src/libmongoc/tests/json-test-operations.c

```diff
@@ -50,6 +50,112 @@ session_from_name (json_test_ctx_t *ctx, const char *session_name)
    }
 }
 
+/* Creates if it does not exist. */
+json_test_worker_thread_t *
+thread_from_name (json_test_ctx_t *ctx, const char *thread_name)
+{
+   json_test_worker_thread_t *wt = NULL;
+   BSON_ASSERT (thread_name);
+
+   if (!strcmp (thread_name, "thread1")) {
+      wt = ctx->worker_threads[0];
+   } else if (!strcmp (thread_name, "thread2")) {
+      wt = ctx->worker_threads[1];
+   } else {
+      test_error ("Unrecognized thread name: %s", thread_name);
+   }
+   if (!wt) {
+      test_error ("Test runner did not create worker threads");
+   }
+   return wt;
+}
+
+json_test_worker_thread_t *
+worker_thread_new (json_test_ctx_t *ctx)
+{
+   json_test_worker_thread_t *wt;
+
+   wt = bson_malloc0 (sizeof (json_test_worker_thread_t));
+   wt->ctx = ctx;
+   /* Note: thread safety of the queue is a bit redundant since the worker
+    * thread locks around it. */
+   wt->queue = q_new ();
+   bson_mutex_init (&wt->mutex);
+   mongoc_cond_init (&wt->cond);
+   return wt;
+}
+
+void
+worker_thread_destroy (json_test_worker_thread_t *wt)
+{
+   q_destroy (wt->queue);
+   bson_mutex_destroy (&wt->mutex);
+   mongoc_cond_destroy (&wt->cond);
+   bson_free (wt);
+}
+
+BSON_THREAD_FUN (json_test_worker_thread_run, wt_void)
+{
+   json_test_worker_thread_t *wt;
+   bson_t *op = NULL;
+   bool shutdown = false;
+
+   wt = wt_void;
+
+   while (!shutdown) {
+      /* Wait a new item in the queue or shutdown. */
+      bson_mutex_lock (&wt->mutex);
+      op = (bson_t *) q_get_nowait (wt->queue);
+      shutdown = wt->shutdown_requested;
+      while (!op && !shutdown) {
+         mongoc_cond_timedwait (&wt->cond, &wt->mutex, 10 * 1000);
+         shutdown = wt->shutdown_requested;
+         op = q_get_nowait (wt->queue);
+      }
+      bson_mutex_unlock (&wt->mutex);
+
+      if (op) {
+         bson_t reply;
+
+         json_test_operation (wt->ctx,
+                              wt->ctx->config->scenario,
+                              op,
+                              wt->ctx->collection,
+                              NULL /* session */,
+                              &reply);
```

Comment:
Ah, `mongoc_client_session_with_transaction` makes sense. The unified test runner shouldn't differ much from transactions (at least not functionally), so I think this would stick around either way.